### PR TITLE
Adding requestId to fallowing apis

### PR DESCRIPTION
**Pull request by Apoorva-GA**
- discover (/patients/discover/carecontexts)
- link (/patients/link)
- create consent request (/consent-requests)
- verify pin (/patients/verify-pin)
- data notification (/health-information​/notification)

### CODE REVIEW

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Clones added
============
- src/main/java/in/projecteka/consentmanager/link/link/LinkRepository.java  3
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestRepository.java  4
- src/main/java/in/projecteka/consentmanager/link/discovery/DiscoveryRepository.java  4
- src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java  2
- src/main/java/in/projecteka/consentmanager/user/TransactionPinRepository.java  2
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

File: postman/NCG.postman_collection.json [hunk 1]
```diff
@@ -270,7 +270,7 @@
 						],
 						"body": {
 							"mode": "raw",
-							"raw": "{\n\t\"hip\": {\n\t\t\"id\": \"10000004\"\n\t}\n}"
+							"raw": "{\n    \"requestId\": \"{{$uuid}}\",\n    \"hip\": {\n        \"id\": \"10000005\"\n    }\n}"
```
Comment (ganesanarun) [inline postman/NCG.postman_collection.json:273]:
```suggestion
							"raw": "{\n    \"requestId\": \"{{$guid}}\",\n    \"hip\": {\n        \"id\": \"10000005\"\n    }\n}"
```
I guess postman does not have $uuid

Comment (ganesanarun) [reply postman/NCG.postman_collection.json:273]: do that everywhere please.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactQueryGenerator.java [hunk 2]
```diff
@@ -22,7 +22,7 @@
             " (consent_request_id, consent_artefact_id, patient_id, consent_artefact, signature, status) VALUES" +
             " ($1, $2, $3, $4, $5, $6)";
     private static final String UPDATE_CONSENT_REQUEST_STATUS_QUERY = "UPDATE consent_request SET status=$1, " +
-            "date_modified=$2 WHERE request_id=$3";
+            "date_modified=$2 WHERE consent_request_id=$3";
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactQueryGenerator.java:25]: we could have said request_id or just use the id column only for tracking request_id. But not a deal breaker.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 3]
```diff
@@ -76,16 +76,19 @@ private static boolean isNotSameRequester(ConsentArtefact consentDetail, String
...
-                .then(validateHIPAndHIU(requestedDetail))
-                .then(saveRequest(requestedDetail, requestId))
-                .then(postConsentRequest.broadcastConsentRequestNotification(ConsentRequest.builder()
-                        .detail(requestedDetail)
-                        .id(requestId)
-                        .build()))
-                .thenReturn(requestId);
+    public Mono<String> askForConsent(RequestedDetail requestedDetail, String requestId) {
+        final String consentRequestId = UUID.randomUUID().toString();
+        return consentRequestRepository.isRequestPresent(requestId)
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java:81]: we can write method to return consent request from repository, do the filter and do rest would be a better approach.

File: postman/NCG.postman_collection.json (hunk 1 above)
Comment (Apoorva-GA) [reply postman/NCG.postman_collection.json:273]: sure

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 4]
```diff
@@ -76,16 +76,19 @@ private static boolean isNotSameRequester(ConsentArtefact consentDetail, String
...
+                .flatMap(requestExists -> (!requestExists)
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java:82]: how about using filter here?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactQueryGenerator.java (hunk 2 above)
Comment (Apoorva-GA) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactQueryGenerator.java:25]: Just thought of keeping it explicit. If thats okay.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 5]
```diff
@@ -76,16 +76,19 @@ private static boolean isNotSameRequester(ConsentArtefact consentDetail, String
...
-        return validatePatient(requestedDetail.getPatient().getId())
-                .then(validateHIPAndHIU(requestedDetail))
-                .then(saveRequest(requestedDetail, requestId))
-                .then(postConsentRequest.broadcastConsentRequestNotification(ConsentRequest.builder()
-                        .detail(requestedDetail)
-                        .id(requestId)
-                        .build()))
-                .thenReturn(requestId);
+    public Mono<String> askForConsent(RequestedDetail requestedDetail, String requestId) {
+        final String consentRequestId = UUID.randomUUID().toString();
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java:80]:
```suggestion
        final String consentRequestId = UUID.randomUUID().toString();
```
Do we need this anymore?

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 6
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java (hunk 5 above)
Comment (Apoorva-GA) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java:80]: Thats the consentRequestId, so we need it. We save it to DB also broadcast it.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java (hunk 5 above)
Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java:80]: I was just wondering why can't we use the unique id sent by the HIU as a consent request id?

File: src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequester.java [hunk 6]
```diff
@@ -34,7 +34,7 @@
         return postDataFlowrequestApproval.broadcastDataFlowRequest(transactionId, dataFlowRequest);
     }
 
-    private Mono<DataFlowRequest> validate(
+    private Mono<DataFlowRequest> saveNotificationRequest(
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequester.java:37]: Is this rename required?

File: src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequester.java [hunk 7]
```diff
@@ -111,10 +111,16 @@ private boolean isValidHIU(String hiuId, ConsentArtefactRepresentation consentAr
...
+        return dataFlowRequestRepository.isRequestPresent(notificationRequest.getRequestId().toString())
+                .flatMap(requestExists -> (!requestExists)
+                        ? saveNotificationRequest(requesterId, notificationRequest)
+                        : Mono.error(ClientError.requestAlreadyExists()));
+    }
+
+    private Mono<Void> saveNotificationRequest(String requesterId, HealthInfoNotificationRequest notificationRequest) {
+        return (!validateRequester(requesterId, notificationRequest))
+            ? Mono.error(ClientError.invalidRequester())
+            : dataFlowRequestRepository.saveNotificationRequest(notificationRequest);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequester.java:123]: Better not mix validation and save opearation.

File: src/main/java/in/projecteka/consentmanager/link/discovery/Discovery.java [hunk 8]
```diff
@@ -41,16 +41,24 @@
...
-                                providerId,
-                                userName,
-                                requestId));
+    public Mono<DiscoveryResponse> patientFor(String userName,
+                                              List<PatientIdentifier> unverifiedIdentifiers,
+                                              String providerId,
+                                              String transactionId,
+                                              String requestId) {
+        return discoveryRepository.isRequestPresent(requestId)
+                .flatMap(requestExists -> (!requestExists)
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/discovery/Discovery.java:50]: Same here, how about filter.

File: src/main/java/in/projecteka/consentmanager/link/discovery/Discovery.java [hunk 9]
```diff
@@ -89,19 +97,20 @@
                 .unverifiedIdentifiers(unverifiedIds)
                 .build();
 
-        var patientRequest = PatientRequest.builder().patient(patient).requestId(requestId).build();
+        var patientRequest = PatientRequest.builder().patient(patient).requestId(transactionId).build();
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/discovery/Discovery.java:100]: This also needs to be renamed?

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/DbOperation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5425033&bid=17791250)
